sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/ProjectTypeCollector.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * SonarSource :: .NET :: Shared library
+ * Copyright (C) 2014-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.dotnet.shared.plugins;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.sonar.api.scanner.ScannerSide;
+
+/**
+ * Collects information about what types of files are in each project (MAIN, TEST, both or none).
+ * The invoker should make sure that no duplicates are added.
+ */
+@ScannerSide
+public class ProjectTypeCollector {
+  private final List<Project> projects = new ArrayList<>();
+
+  void addProjectInfo(boolean hasMainFiles, boolean hasTestFiles) {
+    projects.add(new Project(hasMainFiles, hasTestFiles));
+  }
+
+  Optional<String> getSummary() {
+    int onlyMain = 0;
+    int onlyTest = 0;
+    int mixed = 0;
+    int projectsCount = 0;
+    int noFiles = 0;
+
+    for (Project project : projects) {
+      projectsCount++;
+      if (project.hasMainFiles) {
+        if (project.hasTestFiles) {
+          mixed++;
+        } else {
+          onlyMain++;
+        }
+      } else {
+        // no main files
+        if (project.hasTestFiles) {
+          onlyTest++;
+        } else {
+          noFiles++;
+        }
+      }
+    }
```

Comment:
Isn't this too complicated? I'd say we need `int none = -1, main, test, mixed;` and very simple logic in `addProjectInfo` to increment the correct one.

The `-1` will compensate the top-level module (with appropriate explanation comment).